File: models/store/ItemBatch.js
const { DataTypes } = require('sequelize');
const sequelize = require('../../config/database');

const ItemBatch = sequelize.define('ItemBatch', {
    id: {
        type: DataTypes.UUID,
        primaryKey: true,
        defaultValue: DataTypes.UUIDV4
    },
    item_id: {
        type: DataTypes.UUID,
        allowNull: false,
        references: {
            model: 'items',
            key: 'id'
        }
    },
    institution_id: {
        type: DataTypes.UUID,
        allowNull: true
    },
    batch_number: {
        type: DataTypes.STRING(100),
        allowNull: false,
        unique: true
    },
    quantity: {
        type: DataTypes.INTEGER,
        allowNull: false
    },
    current_quantity: {
        type: DataTypes.INTEGER,
        allowNull: false
    },
    unit_cost: {
        type: DataTypes.DECIMAL(10, 2),
        allowNull: false
    },
    selling_price: {
        type: DataTypes.DECIMAL(10, 2),
        allowNull: true
    },
    expiry_date: {
        type: DataTypes.DATE,
        allowNull: true
    },
    manufacture_date: {
        type: DataTypes.DATE,
        allowNull: true
    },
    supplier_id: {
        type: DataTypes.UUID,
        allowNull: false
    },
    received_date: {
        type: DataTypes.DATE,
        defaultValue: DataTypes.NOW
    },
    status: {
        type: DataTypes.ENUM('active', 'expired', 'depleted', 'recalled'),
        defaultValue: 'active'
    },
    location: {
        type: DataTypes.STRING(100),
        allowNull: true
    },
    notes: {
        type: DataTypes.TEXT,
        allowNull: true
    }
}, {
    tableName: 'item_batches',
    timestamps: true,
});

// Define associations
ItemBatch.associate = (models) => {
    ItemBatch.belongsTo(models.Item, { foreignKey: 'item_id', as: 'item' });
    ItemBatch.belongsTo(models.Supplier, { foreignKey: 'supplier_id', as: 'supplier' });
};

module.exports = ItemBatch;

File: store-service/app/models/item_batch.py
"""ItemBatch model - batch tracking with expiry."""

import uuid
from datetime import datetime

from sqlalchemy import Column, DateTime, Enum, Float, Integer, String, Text
from sqlalchemy.dialects.postgresql import UUID

from app.database import Base
from app.utils.types import BatchStatus


class ItemBatch(Base):
    __tablename__ = "item_batches"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    item_id = Column(UUID(as_uuid=True), nullable=False, index=True)
    institution_id = Column(UUID(as_uuid=True), nullable=True)
    batch_number = Column(String(100), nullable=False, unique=True)
    quantity = Column(Integer, nullable=False)
    current_quantity = Column(Integer, nullable=False)
    unit_cost = Column(Float, nullable=False)
    selling_price = Column(Float, nullable=True)
    expiry_date = Column(DateTime(timezone=True), nullable=True)
    manufacture_date = Column(DateTime(timezone=True), nullable=True)
    supplier_id = Column(UUID(as_uuid=True), nullable=False)
    received_date = Column(DateTime(timezone=True), default=datetime.utcnow)
    status = Column(Enum(BatchStatus), default=BatchStatus.ACTIVE)
    location = Column(String(100), nullable=True)
    notes = Column(Text, nullable=True)

    created_at = Column("createdAt", DateTime(timezone=True), default=datetime.utcnow)
    updated_at = Column("updatedAt", DateTime(timezone=True), default=datetime.utcnow, onupdate=datetime.utcnow)
